internal/parser/parser.go

```diff
@@ -21,6 +24,347 @@ type Scanner interface {
 }
 
 // NewAST parses Flux query and produces an ast.Program.
-func NewAST(src Scanner) (*ast.Program, error) {
-	return nil, errors.New("implement me")
+func NewAST(src Scanner) *ast.Program {
+	p := &parser{s: src}
+	return p.program()
+}
+
+type parser struct {
+	s Scanner
+}
+
+func (p *parser) program() *ast.Program {
+	program := &ast.Program{}
+	for {
+		stmt := p.statement()
+		if stmt == nil {
+			return program
+		}
+		program.Body = append(program.Body, stmt)
+	}
+}
```

Comment:
What do you mean by state or sentinel values? Do you mean the nil being used as a sentinel value?

Reply:
The `eof` token is to accomodate error handling in the AST. It makes it so this can be reused and it was added so block expressions could have better error messages. For example, consider the following:

```
myfunc = (r) => {
    a = r + 1 !
    return a
}
```

If you don't pass the `eof` token, then this would see the `!` and assume that the statement was complete and that there were no more production rules to consider. Then, it would discard the `return` and `a` tokens as both not the expected `}` token. Since we pass the `eof` token in this production rule, it's possible for me to say, "Well, I know I'm going to expect a `}` and that's the only way out, so this must be an invalid statement." I can then add a `BadStatement` node (happening in the error handling PR) and correctly parse the return statement and create a better AST.

This ends up being better for AST construction in my opinion because I'm more likely to get an accurate AST rather than throwing out bad tokens continuously.

Reply:
You can still do that if you do not pass `eof` down to the `statement` production rule. The logic just lives inside the loop instead.

Reply:
Modified this so the `statementList` does not pass down the eof token it is looking for. It now scans the token itself, checks it, and passes the token information down to the production rule to be evaluated.